### DISCUSSION TITLE
[A-L] (12 gennaio 2023) Esercizio 2

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#72] [Esami A-L]
_tengo a precisare che sulla traccia originale ha scritto 2016, si è sbagliato, è del 2023._
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/864aef1d-07f5-4caa-931b-81859ca315e7)

**Comment by Frisbon** [comment]
_sto imparando a farli, potrei sbagliare_
```python
def sudo(A):
    m = len(A)//2

    #controllo subito gli estremi
    if A[0] != A[1]:
        return A[0]
    if A[-1] != A[-2]:
        return A[-1]
    
    # se ho diverso sia a dx che a sx
    if A[m-1] != A[m] and A[m] != A[m+1]: 
        return A[m]

    if A[m-1] != A[m]:
        if m % 2 == 0: return sudo(A[m+1:])
        else:          return sudo(A[:m])
        
    if A[m] != A[m+1]: 
        if m % 2 == 0: return sudo(A[:m])
        else:          return sudo(A[m+1:])
```

**Comment by MarcoLinardi** [comment]
```
 def FindSingleElement(A)
	  i = 0
	  j = len(A)-1
  
	  while i < j
		  mid = (i + j) / 2
		  if mid % 2 == 1
			  mid = mid - 1
  
		  if A[mid] == A[mid+1]
			  i = i + 2
  
		  else
			  j = mid
  
	  return A[i]
```

Sfrutto la caratteristica dell'array di avere tutte coppie tranne un termine, avremo quindi la prima cifra di ogni coppia negli indici pari e la seconda cifra di ogni coppia negli indici dispari.
Se mid è dispari lo decremento di 1 e controllo se A[mid] == A[mid+1], se vero il termine singolo sta nella seconda metà dell'array, se falso nella prima metà, quindi aggiorno gli indici e ricomincia il loop.

**Reply by FeddyLix17** [reply]
l'istruzione i = i +2 non dovrebbe essere invece i = mid +1 ?

**Reply by MarcoLinardi** [reply]
a occhio mi sembra più appropriato addirittura i = mid + 2, credo

**Reply by FeddyLix17** [reply]
cosi però il costo non sarebbe sempre logaritmico

**Comment by IonutCicio** [comment]
L'idea è quella di eseguire una binary search, cambiando la condizione in cui si decide di andare a destra o a sinistra.
Se l'elemento centrale ha un accoppiamento, controllo l'indice più a "sinistra" della coppia:
- se è dispari, vuol dire che tutti gli elementi a sinistra hanno una coppia, e vado a destra
- se è pari, vuol dire che c'è qualche elemento a sinistra che non ha un accoppiamento

```rust
fn search_single_element(array: &[usize]) -> usize {
    if array[0] != array[1] {
        return array[0];
    }

    if array[array.len() - 1] != array[array.len() - 2] {
        return array[array.len() - 1];
    }

    let mut mid = array.len() / 2;

    loop {
        let leftmost = if array[mid] == array[mid + 1] {
            mid
        } else if array[mid - 1] == array[mid] {
            mid - 1
        } else {
            return array[mid];
        };

        mid = if leftmost & 1 == 0 {
            (array.len() + mid) / 2
        } else {
            mid / 2
        }
    }
}
```

**Comment by Elia-Belli** [comment]
## Divide et Impera
**Algoritmo**: con un approccio analogo alla ricerca binaria, controlliamo l'elemento centrale $A[m]$:
- se è singolo, ovvero gli elementi a destra e sinistra sono diversi, abbiamo finito
- altrimenti si procede sull'array dispari risultante dall'analisi, che conterrà sicuramente almeno un elemento singolo

```python
def Alg(A):

	if(len(A) = 1): return A[0]

	m = len(A)/2
	if(A[m] != A[m-1] and A[m] != A[m+1]): return m
	
	if(A[m] == A[m-1]):
		if(m%2 == 0): return Alg(A[0:m-2])	
		else: return Alg(A[m+1:len(A)-1])
	else:
		if(m%2 == 0): return Alg(A[m+2:len(A)-1])	
		else: return Alg(A[0:m-1])

**Comment by daromacarlo** [comment]
```c
def singolo(A : array ordinato e di cardinalità dispari)
	if len(A) == 0{
		print("input sbagliato")
		return null
	}
	if len(A) ==1{
		return A
	}
	m = len(A) / 2
	if A[m] != A[m+1] != A[m-1]{
		return A[m]
	}
	else{
		if A[m] = A[m-1]{
			m -= 1
		}
		if (m % 2) != 1{
			singolo(A[0:m])
		}
		singolo(A[m:len(A)-1])
	}
```

**Comment by niccolo-fato** [comment]
![Screenshot 2025-05-24 alle 21 25 13](https://github.com/user-attachments/assets/833a3f98-6bf9-4787-98a0-96d03bc4aafa)
